core/src/storage2/collections/stash/mod.rs

```diff
@@ -466,6 +466,26 @@ where
         }
     }
 
+    /// Removes the element stored at the given index if any.
+    ///
+    /// This method acts similar to the take API and even still returns an Option.
+    /// However, it guarantees to make no contract storage reads to the indexed
+    /// element and will only write to its internal low-level lazy cache that the
+    /// element at the given index is going to be removed at the end of the contract
+    /// execution.
+    ///
+    /// Calling this method with an index out of bounds for the returns `None` and
+    /// does not `remove` the element, otherwise it returns `Some(())`.
+    pub fn remove(&mut self, at: Index) -> Option<()> {
```

Comment:
After some thinking I decided that we must make this method `unsafe` as well as rename it to `remove_occupied`. The method must trust the caller that it is about to remove an occupied entry. This way we can make it as efficient as it guarantees to be. However, the current implementation forgot to relink the vacant entries that form the doubly linked list. This can be done similarly as it is done in the `take` method. We might want to pull out the relinking code into another method to avoid code duplication.

Reply:
That is fine so long as we provide safe wrappers.

Reply:
> That is fine so long as we provide safe wrappers.

Why would we additionally need safe wrappers? We have safe alternative accessors but also provide this `unsafe` marked accessor in case the caller can be certain about some required invariants. Also this might be useful for other ink! provided data structures to provide more efficient access to themselves. For example ink!'s `storage2::HashMap` could potentially profit here.